### PR TITLE
Clarify registry and client expectations in proposal E

### PR DESCRIPTION
**Pull request by sudo-bmitch**
This pulls up some of the concepts from the requirements into the proposal itself. In particular, the registry section has been expanded with an explanation of garbage collection expectations, and client workflows have been described.

Signed-off-by: Brandon Mitchell <git@bmitch.net>

### CODE REVIEW

File: docs/proposals/PROPOSAL_E.md [hunk 1]
```diff
@@ -184,6 +189,36 @@ For registries that do not support the `referrers` API, a tag MUST be pushed for
 - Periodic garbage collection may be performed by clients pushing new referrers, deleting stale referrers that have been replaced with newer versions, and tags that no longer point to an accessible manifest.
 - Clients can verify the registry does not support the `referrers` API by querying the API and checking for a 404.
 
+#### Client Expectations
+
+##### Creating Artifacts
+
+- For portability, clients should generate artifacts using image-spec until all registries where the artifact is pushed to have been upgraded.
```
Comment (SteveLasker) [inline docs/proposals/PROPOSAL_E.md:196]: This is like saying all image tools should use the docker manifest until all registries support OCI manifests. 
What would be interesting is capturing the behavior for round tripping between a registry that does/doesn't support the new manifest type.
Line 200 captures what I think you're targeting.

File: docs/proposals/PROPOSAL_E.md [hunk 2]
```diff
@@ -94,7 +94,7 @@ Extend the Image Manifest with a refers field (existing registries should ignore
 }
 ```
 
-### Registry HTTP API
+### Registry Changes
```
Comment (sajayantony) [inline docs/proposals/PROPOSAL_E.md:97]: Similar to Image spec maybe reword this to Distribution spec?

```suggestion
### Distribution Spec
```

File: docs/proposals/PROPOSAL_E.md [hunk 3]
```diff
@@ -184,6 +189,36 @@ For registries that do not support the `referrers` API, a tag MUST be pushed for
...
+- Other metadata should be included in the manifest annotations.
+- When pushing, the registry should be checked for the referrers API support.
+- If the referrers API is not available, the artifact should be pushed with a tag using the digest schema. Tags with the digest schema should not be pushed if the referrers API is available.
+
+##### Pulling Artifacts
+
+To pull artifacts that reference an existing manifest:
+
+- Clients requests artifacts associated with a manifest first query the referrers API using the digest of the requested manifest.
+- If the request fails, clients must fall back to listing tags in the repository and pull manifests for each tag with the matching `<alg>` and `<ref>` prefix.
```
Comment (sajayantony) [inline docs/proposals/PROPOSAL_E.md:207]: Can't clients avoid the tag listing API if they know the digest-type of the tag they need ?
Since we have this capabilty with tags I would say it would be worth considering if we can add the artifactType filter to the API as well.

File: docs/proposals/PROPOSAL_E.md [hunk 4]
```diff
@@ -184,6 +189,36 @@ For registries that do not support the `referrers` API, a tag MUST be pushed for
...
+- Clients then pull any artifact matching their criteria from the API response or the list generated from the tag query.
+
+##### Copying Images
+
+- Tooling that copies images between registries should recursively query for referrers.
```
Comment (sajayantony) [inline docs/proposals/PROPOSAL_E.md:212]:
```suggestion
- Tooling that copies images between registries may recursively query for referrers and copy them.
```

File: docs/proposals/PROPOSAL_E.md [hunk 5]
```diff
@@ -184,6 +189,36 @@ For registries that do not support the `referrers` API, a tag MUST be pushed for
...
+- Copying an artifact that refers to an image uses the same pull and push steps described above and should support both existing registries and registries with the `referrers` API on both the pull and push.
+
+##### Deleting Manifests
+
+For managing existing registries without the `referrers` API:
+
+- Client tooling that deletes manifests should also delete digest tags that reference that manifest.
+- Client tooling should periodically check for dangling digest tags that refer to missing manifest, and prune those tags.
```
Comment (sajayantony) [inline docs/proposals/PROPOSAL_E.md:220]: It might be worth considering the scenario where folks want to keep some referrers without dropping them. This prune recommendation might mean they should tag differently to keep things around.

File: docs/proposals/PROPOSAL_E.md (hunk 1 above)
Comment (sudo-bmitch) [reply docs/proposals/PROPOSAL_E.md:196]: Line 200 handles what to do if the new API is not available. But if you push an unknown manifest to the registry, it will reject it from any API. If the client knows how to round trip the artifact, then why not push the artifact as an image-spec and skip the conversion and digest change on the content? Do that round tripping process need to be standardized so an artifact changed by one tool is readable by another? I was trying to avoid getting into that complexity because it feels like a time sink that would distract us from getting this finished.

File: docs/proposals/PROPOSAL_E.md (hunk 3 above)
Comment (sudo-bmitch) [reply docs/proposals/PROPOSAL_E.md:207]: If clients know the digest of the artifact (the value of `<hash>`), they can skip the tag listing, but I don't know how they'd get that.

File: docs/proposals/PROPOSAL_E.md (hunk 5 above)
Comment (sudo-bmitch) [reply docs/proposals/PROPOSAL_E.md:220]: I'm leaning towards keeping the scope in the proposal to what's needed to manage an artifact that refers to another manifest. The fear is if we start documenting some other scenarios, we'll be asked to document them all. But to your point, yes, an artifact can be tagged and pulled directly.

Comment (imjasonh) [reply docs/proposals/PROPOSAL_E.md:220]: Can I ask for the "should" to be changed to "could" or "may"? As we start to think about how this will land as official spec changes, words like "SHOULD" get new meaning, and I'd like to avoid recommending with the force that a "SHOULD" brings with it.

Comment (sudo-bmitch) [reply docs/proposals/PROPOSAL_E.md]: @imjasonh Done.

File: docs/proposals/PROPOSAL_E.md [hunk 6]
```diff
@@ -94,7 +94,7 @@ Extend the Image Manifest with a refers field (existing registries should ignore
 }
 ```
 
-### Registry HTTP API
+### Distribution Spec
```
Comment (jdolitsky) [inline docs/proposals/PROPOSAL_E.md:97]: this is changing of the template format, but at this point I think its ok

Comment (sajayantony) [reply docs/proposals/PROPOSAL_E.md:97]: Oh ok. That wasn't my intention. I think we can ignore this comment then.

**Review by jdolitsky — APPROVED** [review]
